crates/red_knot_python_semantic/resources/mdtest/invalid_syntax.md

```diff
@@ -26,13 +26,10 @@ def True(for):
     # error: [invalid-syntax]
     pass
 
-# TODO: Why is there two diagnostics for the same error?
-
 # error: [invalid-syntax]
 # error: [invalid-syntax]
 # error: [invalid-syntax]
```

Comment:
Why are there still three invalid-syntax errors? Can we add messages clarifying what the three errors are? I would expect two (one for `while` used as identifier, the other for `pass` used as identifier.)

Reply:
In brief, our parser cannot recover from certain syntax errors completely. For example, if there's a keyword (like `for`) in the middle of another statement (like function definition), then it's more likely that the rest of the tokens are going to be part of the `for` statement and not the function definition. But, it's not necessary that the remaining tokens are valid in the context of a `for` statement. This is what's happening here. There's a function definition before this part of the code that the parser doesn't recover well from:

```py
def True(for):
    pass

for while in pass:
    pass
```

Try commenting out the function definition and you'll get the expected number of errors. I could add this as a general comment to this file.

Playground: https://play.ruff.rs/a36e1a68-abd9-404d-bb79-146855c6d521